core/services/relay/evm/write_target.go

```diff
@@ -69,7 +84,208 @@ func NewWriteTarget(ctx context.Context, relayer *Relayer, chain legacyevm.Chain
+	}
+
+	if hex.EncodeToString(reportMetadata.WorkflowCID[:]) != rawRequest.Metadata.WorkflowID {
+		return "", fmt.Errorf("WorkflowID in the report does not match WorkflowID in the request metadata. Report WorkflowID: %+v, request WorkflowID: %+v", reportMetadata.WorkflowCID, rawRequest.Metadata.WorkflowID)
+	}
+
+	if !bytes.Equal(reportMetadata.ReportID[:], r.Inputs.SignedReport.ID) {
+		return "", fmt.Errorf("ReportID in the report does not match ReportID in the inputs. reportMetadata.ReportID: %x, Inputs.SignedReport.ID: %x", reportMetadata.ReportID, r.Inputs.SignedReport.ID)
+	}
+
+	return r.Config.Address, nil
+}
+
+func decodeReportMetadata(data []byte) (metadata ReportV1Metadata, err error) {
+	if len(data) < metadata.Length() {
+		return metadata, fmt.Errorf("data too short: %d bytes", len(data))
+	}
+	return metadata, binary.Read(bytes.NewReader(data[:metadata.Length()]), binary.BigEndian, &metadata)
+}
+
+func (rm ReportV1Metadata) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.BigEndian, rm)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (rm ReportV1Metadata) Length() int {
+	bytes, err := rm.Encode()
+	if err != nil {
+		return 0
+	}
+	return len(bytes)
+}
+
+// Note: This should be a shared type that the OCR3 package validates as well
+type ReportV1Metadata struct {
+	Version             uint8
+	WorkflowExecutionID [32]byte
+	Timestamp           uint32
+	DonID               uint32
+	DonConfigVersion    uint32
+	WorkflowCID         [32]byte
+	WorkflowName        [10]byte
+	WorkflowOwner       [20]byte
+	ReportID            [2]byte
+}
+
+func getChainInfo(chainID uint64) (monitor.ChainInfo, error) {
+	chainSelector := chainselectors.EvmChainIdToChainSelector()[chainID]
+	chainFamily, err := chainselectors.GetSelectorFamily(chainSelector)
+	if err != nil {
+		return monitor.ChainInfo{}, fmt.Errorf("failed to get chain family for selector %d: %w", chainSelector, err)
+	}
+	chainDetails, err := chainselectors.GetChainDetailsByChainIDAndFamily(strconv.Itoa(int(chainID)), chainFamily)
+	if err != nil {
+		return monitor.ChainInfo{}, fmt.Errorf("failed to get chain details for chain %d and family %s: %w", chainID, chainFamily, err)
+	}
+
+	neworkName, err := getNetworkName(chainDetails.ChainName)
+	if err != nil {
+		return monitor.ChainInfo{}, fmt.Errorf("failed to get network name for chain %d: %w", chainID, err)
+	}
+
+	return monitor.ChainInfo{
+		ChainFamilyName: chainFamily,
+		ChainID:         strconv.Itoa(int(chainID)),
+		NetworkName:     neworkName,
+		NetworkNameFull: chainDetails.ChainName,
+	}, nil
+}
+
+func getNetworkName(fullName string) (string, error) {
+	parts := strings.Split(fullName, "-")
+	if len(parts) < 2 {
+		return "", fmt.Errorf("invalid network name format: %s", fullName)
+	}
+	name := parts[1]
+	if !slices.Contains([]string{"testnet", "mainnet", "devnet"}, name) {
```

Comment:
Can you please write some tests, this impl doesn't look correct to me.

Reply:
How about using a regex?

```
func extractNetwork(selector string) string {
	// Create a regexp pattern that matches any of the three.
	re := regexp.MustCompile(`(mainnet|testnet|devnet)`)
	return re.FindString(selector)
}
```

Reply:
Updated it to use regex and added some testcases.